core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala

```diff
@@ -0,0 +1,134 @@
+ * different taskSets can learn experience from other taskSet to avoid allocating tasks on
+ * problematic executors.
+ */
+class SimpleStrategy(
+    maxFailureTaskNumber: Int,
+    maxBlackExecutorNumber: Int,
+    val expireTimeInMillisecond: Long
+  )extends BlacklistStrategy {
+
+  private def getSelectedExecutorMap(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus]) = {
+    executorIdToFailureStatus.filter{
+      case (id, failureStatus) => failureStatus.totalNumFailures > maxFailureTaskNumber
+    }
+  }
+
+  // As this is a taskId unrelated strategy, the input taskId will be ignored
+  def getExecutorBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus],
+      taskId: Long): Set[String] = {
+    getSelectedExecutorMap(executorIdToFailureStatus).keys.toSet
+  }
+
+  def getNodeBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus]): Set[String] = {
+    getSelectedExecutorMap(executorIdToFailureStatus)
+      .groupBy{case (id, failureStatus) => failureStatus.host}
+      .filter {case (host, executorIdToFailureStatus) =>
+        executorIdToFailureStatus.size > maxBlackExecutorNumber}
+      .keys.toSet
+  }
+}
+
+/**
+ * This strategy is applied as default to keep the same semantics as original. It's an taskId
+ * related strategy. If an executor failed running "task A", then we think this executor is
+ * blacked for "task A". And we think the executor is still healthy for other task. node blacklist
+ * is always empty.
+ *
+ * It was the standard behavior before spark 1.6
+ */
+class DefaultStrategy(val expireTimeInMillisecond: Long) extends BlacklistStrategy {
+  def getExecutorBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus],
+      taskId: Long): Set[String] = {
+    executorIdToFailureStatus.filter{
+      case (_, failureStatus) => failureStatus.numFailuresPerTask.keySet.contains(taskId)
+    }.keys.toSet
+  }
+
+  def getNodeBlacklist(
+      executorIdToFailureStatus: mutable.HashMap[String, FailureStatus]): Set[String] =
+        Set.empty[String]
+}
+
+/**
+ * Create BlacklistStrategy instance according to SparkConf
+ */
+object BlacklistStrategy {
+  def apply(sparkConf: SparkConf): BlacklistStrategy = {
+    val timeout = sparkConf.getLong("spark.scheduler.executorTaskBlacklistTime", 0L)
+    sparkConf.get("spark.scheduler.blacklist.strategy", "default") match {
+      case "default" =>
+        new DefaultStrategy(timeout)
+      case "threshold" =>
+        new SimpleStrategy(
+            sparkConf.getInt("spark.scheduler.blacklist.threshold.maxFailureTaskNumber", 3),
+            sparkConf.getInt("spark.scheduler.blacklist.threshold.maxBlackExecutorNumber", 3),
+            timeout)
+      case "strict" =>
+        // A special case of SimpleStrategy: Once task failed  at executor,
+        // put the executor and its node into blacklist.
+        new SimpleStrategy(0, 0, timeout)
```

Comment:
I don't really see the point of adding "strict" -- do you think it will be a very common use case?  Otherwise its just another thing to document and might confuse users a bit.

I'd also like to think of better names for the other strategies as well.  It seems the key difference is whether or not they are task-specific, which isn't very clear from "default" and "threshold".  I'm terrible at naming myself ... something like "singleTask" and "completeExecutor"?